tools/autograd/gen_inplace_or_view_type.py

```diff
@@ -63,16 +63,76 @@
 for key in VIEW_FUNCTIONS_WITH_METADATA_CHANGE:
     VIEW_FUNCTIONS[key] = 'self'
 
+
+# The full set of composite view ops, including overloads.
+COMPOSITE_VIEW_FULL_NAMES = set([
+    'chunk',
+    'contiguous',
+    'reshape',
+    'reshape_as',
+    'expand_as',
+    'view_as',
+    'real',
+    'imag',
+    'narrow',
+    'narrow.Tensor',
+    'movedim.intlist',
+    'movedim.int',
+    'moveaxis.intlist',
+    'moveaxis.int',
+    'tensor_split.sections',
+    'tensor_split.indices',
+    'tensor_split.tensor_indices_or_sections',
+    'swapdims',
+    'swapaxes',
+    'diagonal.Dimname',
+    'flatten.using_ints',
+    'flatten.named_out_dim',
+    'flatten.using_names',
+    'flatten.DimnameList',
+    'select.Dimname',
+    'squeeze.dimname',
+    'transpose.Dimname',
+    'unbind.Dimname',
+    'numpy_T',
+    # inplace-view composites
+    'squeeze_.dimname',
+    'swapaxes_',
+    'swapdims_',
+])
+
+# This is the set of composite view ops currently used by gen_inplace_or_view_type.py
+# TODO: the existing list of composite ops used by gen_inplace_or_view_type.py is wrong
+# We should rationalize these differences.
+COMPOSITE_VIEWS = set([x.partition('.')[0] for x in COMPOSITE_VIEW_FULL_NAMES]) - set([
+    # Not sure why these aren't marked as composite?
+    'moveaxis',
+    'numpy_T',
+    # inplace-view composites aren't included either... should they be?
+    'squeeze_.dimname',
+    'swapaxes_',
+    'swapdims_',
+    # For everything below here, there are multiple overloads of the listed op, where one is composite and the other is not.
+    # the list used by gen_inplace_or_viewtype can't use them because it only uses the base name of the op
+    # so it can't distinguish between the overloads.
+    # Maybe this codegen should just use the full base.overload name?
+    'select',
+    'diagonal',
+    'transpose',
+    'unbind',
+    'flatten',
+    'squeeze',
+]) | set([
+    # detach is listed as composite, even though it's not... it's CompositeExplicitAutograd.
+    'detach'
+])
```

Comment:
Probably out of scope for this PR, but these decls seem like the kind of thing that could be served by the tagging machinery @ezyang and @Chillee have been talking about... that would localize the classification in `native_functions` at least, though you'd still need to know the taxonomy when adding new ops that fit into these categories.

Reply:
I originally blindly took the hardcoded list from this file and modified it to fit my needs, but I think this information is already fully inspectable from native_functions.yaml?

To get all view ops: Look for schemas that have an immutable, aliased output (e.g. `-> Tensor(a)`). We already depend on all views having this type of schema in the CPUFallback, so this seems pretty safe.

To get all composite views: Filter above by NativeFunctions with a CompositeImplicitAutograd kernel.

I have a feeling this is going to give me some extra views outside of the hardcoded list above, but I'm going to try that instead, and hopefully get rid of any changes to this file.

I'm still not sure why `gen_inplace_or_view_type.py` needs a hardcoded list in the first place (which is also wrong - @albanD do you happen to know why?) - but I'll leave that out of scope.

Reply:
yeah, doing this gave me 3 extra view ops:
- `_fw_primal`
- `crow_indices` (used in SparseCSR)
- `col_indices` (used in SparseCSR)

functionalization currently doesn't work with sparse anyway, and I see that `_fw_primal()` is currently not implemented (the impl in `VariableMethodStubs.cpp` throws an exception), so I think it seems fine to use this method to find view ops instead of the hardcoded list.

Reply:
one annoying thing is that it's harder to automatically detect inplace + mutation ops like `transpose_` from the codegen :(. I can probably do it by lopping off the trailing underscore, looking up the corresponding NativeFunction based on the operator name and checking it's schema to see if it's a view.

Feels a little convoluted, but I don't think that would ever be incorrect (and is less heavyweight than maintaining actual tags for it in the yaml)

Reply:
Yeah I think that's a solid rationale - I guess my rule of thumb would be that if the pattern matching rules are robust enough, i.e. if they check properties that clearly follow from the categories you're matching into and not just based on coincidentally occurring characteristics, *and* the relationships between categories are simple - like, ideally nonoverlapping (every op matches at most one category) - then it's probably a win. The double hop of the transpose situation is probably fine, but might warrant some defensiveness in the introspection code - like, if you find a trailing-underscore version of an op but no underscoreless version, don't just fail silently but actually throw something, that kind of thing.

Reply:
I talked to @ezyang about using tags for just the view + inplace ops, which is pretty lightweight to add and seems directionally where we want to go anyway (and wouldn't require me to construct another Dictionary[OperatorName, NativeFunction] somewhere in the codegen)

Reply:
The list of views in `gen_inplace_or_view_type.py` should be an up to date list of the elementary autograd functions that are views.
`_fw_primal` is manually implemented in VariableTypeManual.
crow and col have been fixed in master no? I recall a PR about it a while back.